tracing-attributes/src/expand.rs

```diff
@@ -191,8 +191,14 @@ fn gen_block<B: ToTokens>(
     })();
 
     let err_event = match args.err_mode {
-        Some(ErrorMode::Display) => Some(quote!(tracing::error!(error = %e))),
-        Some(ErrorMode::Debug) => Some(quote!(tracing::error!(error = ?e))),
+        Some(FormatMode::Display) => Some(quote!(tracing::error!(error = %e))),
+        Some(FormatMode::Debug) => Some(quote!(tracing::error!(error = ?e))),
+        _ => None,
+    };
+
+    let ret_event = match args.ret_mode {
+        Some(FormatMode::Display) => Some(quote!(tracing::trace!(return = %#block))),
+        Some(FormatMode::Debug) => Some(quote!(tracing::trace!(return = ?#block))),
```

Comment:
actually evaluating the block _inside_ of the tracing macro (`return = #block`) means the block will _only_ be evaluated if the `TRACE` level is enabled...which seems quite wrong, as it means the execution of the function might depend on the tracing filtering configuration. instead, we should _always_ evaluate the block outside of the macro, binding the result to a variable, and then emit the event with that value.

in the case where the `err` attribute is also enabled, that would mean moving the return event into the `Ok` match arm, and recording the value inside the `Ok`. when `err` is not enabled, it would look more like:
```suggestion
        quote!(
            let ret = #block;
            tracing::trace!(return = %ret))),
            ret
        )
```
or similar.

Reply:
also, I'm not sure if I like that we'll always emit the return event at the `TRACE` level. right now, there's no way to change the level of the return value event, so users could write code like 
```rust
#[instrument(level = "info", ret(Debug))]
fn do_something(a: A) -> B {
    // ...
}
```
enable the `INFO` level, and be surprised when the span is enabled but the return value event is not emitted. instead, I think the best thing to do might be to emit the return value event at the _same_ level as the span generated by `instrument`. that seems like a reasonable default behavior to me.

it's certainly possible that people might want to emit the return value event at a _different_ level, but i think that should be done by allowing users to explicitly set the level for the return event. i wouldn't implement that now, though, as it requires some extra work --- instead, I'd make a note of that as a potential follow-up if there's a need for it.

Reply:
> actually evaluating the block inside of the tracing macro (return = #block) means the block will only be evaluated if the TRACE level is enabled...which seems quite wrong, as it means the execution of the function might depend on the tracing filtering configuration. instead, we should always evaluate the block outside of the macro, binding the result to a variable, and then emit the event with that value.

I guess this is fixed by f7cd4c476d6aaaf3a5f510e6d6f4db77a8afb7cc.

> enable the INFO level, and be surprised when the span is enabled but the return value event is not emitted. instead, I think the best thing to do might be to emit the return value event at the same level as the span generated by instrument. that seems like a reasonable default behavior to me.

That sounds really good to me. This looks achievable by using `InstrumentArgs::level()`.